Added new scenario to fill up a given volumen

agullon opened the pull request:
### Description
Scenario to fill up a given PersistenVolumeClaim by creating a temp file of a given size from a pod that has the volume mounted in the given path. The purpose of this scenario is to fill up a volume to understand faults cause by the application using this volume.

paigerube14 commented on config/config.yaml:
```diff
@@ -39,6 +39,8 @@ kraken:
             - scenarios/zone_outage.yaml
         -   application_outages:
             - scenarios/app_outage.yaml
+        -   pvc_scenarios:
```
Can you also add this to the config_performance.yaml file

agullon replied:
done

paigerube14 commented on docs/pvc_scenario.md:
````diff
@@ -0,0 +1,21 @@
+
+##### Sample scenario config
+```
+pvc_scenario:                        # Scenario to stop all the nodes for specified duration and restart the nodes
````
Comment on this line does not match the scenario

agullon replied:
done

chaitanyaenr commented:
Hi @agullon, we might want to run the pre-commit using the instructions: https://github.com/cloud-bulldozer/kraken/blob/master/docs/contribute.md#fix-formatting to fix the CI failures - mostly cosmetic. We can run $ pre-commit run --all-files and git add, commit to auto fix them.

chaitanyaenr commented on kraken/pvc/pvc_scenario.py:
```diff
@@ -0,0 +1,57 @@
+                config_yaml = yaml.full_load(f)
+                scenario_config = config_yaml["pvc_scenario"]
+                pod_name = scenario_config.get("pod_name", "")
+                namespace = scenario_config.get("namespace", "")
```
Maybe exit if namespace/podname is empty? Thoughts?

agullon replied:
good idea, it's done

paigerube14 commented:
This scenario requires a pod to already be created with maps to a PVC, correct? It might be nice to be able to nice to include the pod creation as part of this setup/scenario because the user might not have one set up. I think we should do all of the set up of each scenario within kraken and not require any set up before the scenario runs

Thoughts @chaitanyaenr @agullon

chaitanyaenr commented on docs/pvc_scenario.md:
````diff
@@ -0,0 +1,21 @@
+##### Sample scenario config
+```
+pvc_scenario:                        # Scenario to stop all the nodes for specified duration and restart the nodes
+  pod_name: pod_name                 # Name of the pod with the PVC linked to
````
Maybe leverage pod labels to grab the pod-name to ease the CI/automated use cases where we might not know the pod name especially when it's appended with a random id each time it gets created. Something along the lines of user providing the pod-label to query and the number of pods to pick for injecting the chaos. Thoughts?

chaitanyaenr commented:
> This scenario requires a pod to already be created with maps to a PVC, correct? It might be nice to be able to nice to include the pod creation as part of this setup/scenario because the user might not have one set up. I think we should do all of the set up of each scenario within kraken and not require any set up before the scenario runs
> 
> Thoughts @chaitanyaenr @agullon

@paigerube14 There might be cases where applications running will have PV's allocated in public clouds in which case users might have a use case to target that particular PV to understand how it impacts that particular application. For example, kafka running on AWS will have a PV allocated and @agullon is looking at filling up that particular PV to understand the behavior of Kafka I guess. 

Maybe we can add a note about only targeting the pods with existing PV's for this scenario? Thoughts?

agullon commented:
@paigerube14 @chaitanyaenr I agree with the idea that we shouldn't do any prior set up before running the scenario. My first idea was to pass the PVC name as a parameter, but I pass the name of the namespace and pod instead. The reason I did it this way is because the implementation is easier. I'll investigate if we can get the pod, namespace and mount path from the PVC name (maybe we need to create a temp pod to do that).

What do you think about this approach?

chaitanyaenr commented:
> @paigerube14 @chaitanyaenr I agree with the idea that we shouldn't do any prior set up before running the scenario. My first idea was to pass the PVC name as a parameter, but I pass the name of the namespace and pod instead. The reason I did it this way is because the implementation is easier. I'll investigate if we can get the pod, namespace and mount path from the PVC name (maybe we need to create a temp pod to do that).
> 
> What do you think about this approach?



> @paigerube14 @chaitanyaenr I agree with the idea that we shouldn't do any prior set up before running the scenario. My first idea was to pass the PVC name as a parameter, but I pass the name of the namespace and pod instead. The reason I did it this way is because the implementation is easier. I'll investigate if we can get the pod, namespace and mount path from the PVC name (maybe we need to create a temp pod to do that).
> 
> What do you think about this approach?

Maybe we can get the pod using the label and namespace specified in the config and then rsh/exec to the pod like the current implementation to create the file under /tmp to use the space if the mount_path is not defined or under the mount_path if specified with an assumption that the pod label/namespace specified by the user has a PV attached to it. Thoughts?

agullon commented:
I changed the approach. Now, the `pod_name` and `mount_path` are retrieved from the `pvc_name` and the `namespace`. Also added a new input parameter `fill_percentage` to specify how much the PVC will be filled up during the test.

chaitanyaenr commented on kraken/pvc/pvc_scenario.py:
```diff
@@ -0,0 +1,139 @@
+                    sys.exit(1)
+
+                # Get pod name
+                command = "oc describe pvc %s -n %s | grep \"Used By:\" | grep -Eo '[^: ]*$'" % (
```
Maybe s/oc/kubectl to be able to run the scenario on kubernetes environments as well? Thoughts?

agullon replied:
Done. I change the `oc` commands to `kubectl` commands.

chaitanyaenr commented:
@agullon hitting the following error when running the scenario:

2021-11-22 20:07:19,825 [INFO] Executing scenarios for iteration 0
2021-11-22 20:07:19,825 [INFO] connection set up
127.0.0.1 - - [22/Nov/2021 20:07:19] "GET / HTTP/1.1" 200 -
2021-11-22 20:07:19,827 [INFO] response RUN
2021-11-22 20:07:19,827 [INFO] Running PVC scenario
Traceback (most recent call last):
  File "run_kraken.py", line 288, in <module>
    main(options.cfg)
  File "run_kraken.py", line 218, in main
    pvc_scenario.run(scenarios_list, config)
  File "/root/pvc/kraken/kraken/pvc/pvc_scenario.py", line 55, in run
    volume_mounts_list_json = json.loads(volume_mounts_list)
  File "/usr/lib64/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python3.6/json/decoder.py", line 342, in decode
    raise JSONDecodeError("Extra data", s, end)
json.decoder.JSONDecodeError: Extra data: line 72 column 1 (char 1842)

agullon commented:
> @agullon hitting the following error when running the scenario:
> 
> ```
> 2021-11-22 20:07:19,825 [INFO] Executing scenarios for iteration 0
> 2021-11-22 20:07:19,825 [INFO] connection set up
> 127.0.0.1 - - [22/Nov/2021 20:07:19] "GET / HTTP/1.1" 200 -
> 2021-11-22 20:07:19,827 [INFO] response RUN
> 2021-11-22 20:07:19,827 [INFO] Running PVC scenario
> Traceback (most recent call last):
>   File "run_kraken.py", line 288, in <module>
>     main(options.cfg)
>   File "run_kraken.py", line 218, in main
>     pvc_scenario.run(scenarios_list, config)
>   File "/root/pvc/kraken/kraken/pvc/pvc_scenario.py", line 55, in run
>     volume_mounts_list_json = json.loads(volume_mounts_list)
>   File "/usr/lib64/python3.6/json/__init__.py", line 354, in loads
>     return _default_decoder.decode(s)
>   File "/usr/lib64/python3.6/json/decoder.py", line 342, in decode
>     raise JSONDecodeError("Extra data", s, end)
> json.decoder.JSONDecodeError: Extra data: line 72 column 1 (char 1842)
> ```
> 
> Here's the scenario file:
> 
> ```
> pvc_scenario:
>   pvc_name: prometheus-k8s-db-prometheus-k8s-0
>   namespace: openshift-monitoring   # Namespace where the PVC is
>   fill_percentage: 50           # Target percentage to fill up the cluster, value must be higher than current percentage, valid values are between 0 and 99
>   duration: 60                  # Duration in seconds for the fault
> ```

Thanks @chaitanyaenr , I fix it. Now I look for the volume name to get the mount path. I also get the container name to secure the execution of the commands in the container's pods.

paigerube14 commented:
Can we add in a sample yaml to create a pod and pvc for this scenario? will be useful to have when adding a CI test for this scenario

agullon commented:
> Can we add in a sample yaml to create a pod and pvc for this scenario? will be useful to have when adding a CI test for this scenario

Sure, I'll work on this.

agullon commented:
Hi @paigerube14 I added a yaml file to create a temp pv, pvc and pod to run the pvc scenario for CI, please, take a look at the commit https://github.com/cloud-bulldozer/kraken/pull/182/commits/db2fdfeb814bf8a46adb5d64336fa988a7cec080

paigerube14 commented:
@agullon thanks so much! Will use that in my testing today. Sorry for the delay

agullon commented:
@chaitanyaenr @paigerube14 Added support to find the PVC from a given pod on commit https://github.com/cloud-bulldozer/kraken/pull/182/commits/533369cc9b0049a0606536dd63296116fabaf5e4

chaitanyaenr reviewed (approved):
LGTM. Great addition to the supported chaos scenarios @agullon!

paigerube14 commented on kraken/pvc/pvc_scenario.py:
```diff
@@ -0,0 +1,175 @@
+                        volume_name = entry["name"]
+                        pvc_name = entry["persistentVolumeClaim"]["claimName"] if (len(pvc_name) == 0) else pvc_name
+                        break
+                logging.info("Volumen name: %s" % volume_name)
```
small edit, take out the "n" at the end of volume

agullon replied:
done

chaitanyaenr commented:
@agullon Thoughts on adding support for this scenario in Kraken-hub post merge - https://github.com/cloud-bulldozer/kraken-hub/blob/main/docs/test_your_changes.md#adding-a-new-scenario?

paigerube14 reviewed (approved):
Used the volume_sceanrio.yaml to create a pv, pvc, and pod to use in the scenario and it worked great! LGTM

paigerube14 commented on CI/scenarios/volume_scenario.yaml:
```diff
@@ -0,0 +1,44 @@
+    - name: kraken-test-container
+      image: 'image-registry.openshift-image-registry.svc:5000/openshift/httpd:latest'
+      volumeMounts:
+        - mountPath: "/home/krake-dir/"
```
sorry, one more thing. the directory is missing the end "n" in kraken

chaitanyaenr replied:
Merged the PR when this comment was in flight, my bad.

paigerube14 replied:
No problem!

agullon commented:
> @agullon Thoughts on adding support for this scenario in Kraken-hub post merge - https://github.com/cloud-bulldozer/kraken-hub/blob/main/docs/test_your_changes.md#adding-a-new-scenario?

Yes, I will work on it during the next days.
